aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1252,14 +1252,20 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
     at::native::_baddbmm_mkl_(self_or_result, batch1, batch2, beta, alpha);
   } else { // split along batch dimension
     if (is_bmm_out) {
-      for (int64_t b = 0; b < bs; b++) {
-        auto r = self_or_result.select(0, b);
-        native::mm_cpu_out(batch1.select(0, b), batch2.select(0, b), r);
-      }
+      auto bmm_out_fn = [&](uint64_t start, uint64_t end) {
+        for (int64_t b = start; b < end; b++) {
+          auto r = self_or_result.select(0, b);
```

Comment:
Yea at::parallel_for should operate on data_ptr instead of tensors. I guess the best way to fix here is to make sure the tensor is contiguous and work on data_ptr directly. 
As suggested by @ezyang adding NoGradMode might help in many cases, but it won't be able to propagate InferenceMode TLS properly, so you'd expect some weird errors when using this operator inside InferenceMode(which I think is enabled on mobile for most use cases).

Reply:
@ailzhang so to make sure I understand. Reason Tensor ops are not "allowed" inside at::parallel is because we are not able to propagate various TLS state through at::parallel and thus have those child invocations inherit the state from parent. Is my understanding correct that we want to preserve TLS state across threads for autograd (and pardon me if my understandign is not correct)? Or we also want to propagate other state such as RecordFunction? If we want to preserve state besides autograd then I would want to understand if it is necessary. Else this prevents us from doing lot of intra op parallelization.
Also in this particular instance just making sure that we are not in autograd mode be sufficient?